### PR TITLE
Windows: Fix deleting directory in JNI code

### PR DESCRIPTION
**Pull request by meteorcloudy**
`bazel clean` command sometimes fails on Windows with
"Directory not empty" error. But when checking the directory after the
failure, we don't see any file or directory under it.
After debugging, we cannot find any unclosed file handle, either.

However, when trying to delete the bazel output directory with `rm -rf" it
succeeds everytime. The reason is that Cygwin worked around a Windows
issue at this commit:
https://github.com/Alexpux/Cygwin/commit/28fa2a72f810670a0562ea061461552840f5eb70

The problem is, I quote:
""
Intensive testing shows that sometimes directories, for which
the delete disposition has already been set, and the deleting
handle is already closed, can linger in the parent dir for a
couple of ms for no apparent reason (Windows Defender or other
real-time scanners are suspect).
""

In this commit, we use a similar way to retry deleting the directory when
its "deleted" sub-dirs are still lingering under it.

Related https://github.com/bazelbuild/bazel/issues/5907

### CODE REVIEW

**Comment by meteorcloudy** [comment]
While this PR fix `bazel clean` command, but it doesn't fix the Java builder failure in https://github.com/bazelbuild/bazel/issues/5907, because the Java builder uses a different way to delete directory. I still need to find a fix for that.

File: src/main/native/windows/file.cc [hunk 1]
```diff
@@ -446,20 +446,97 @@ int DeletePath(const wstring& path, wstring* error) {
...
+          static const wstring kDot(L".");
+          static const wstring kDotDot(L"..");
+          bool found_valid_file = false;
+          bool found_pending_delete_file = false;
+          WIN32_FIND_DATAW metadata;
+          HANDLE handle = ::FindFirstFileW((winpath + L"\\*").c_str(), &metadata);
+          if (handle != INVALID_HANDLE_VALUE) {
+            do {
+              if (kDot != metadata.cFileName && kDotDot != metadata.cFileName) {
+                std::wstring wname = winpath + L"\\" + metadata.cFileName;
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: `wname` looks confusingly similar to `wpath` and `winpath`.
How about calling it `child` instead?

File: src/main/native/windows/file.cc [hunk 2]
```diff
@@ -446,20 +446,97 @@ int DeletePath(const wstring& path, wstring* error) {
...
+      // Sometimes a deleted directory lingers in its parent dir
+      // after the deleting handle has already been closed.
+      // In this case we check the content of the parent directory,
+      // if we don't find any valid file, we try to delete it again after 5 ms.
+      // But we don't want to hang infinitely because another application
+      // can hold the handle for a long time. So try at most 20 times,
+      // which means a process time of 100-120ms.
+      // Inspired by https://github.com/Alexpux/Cygwin/commit/28fa2a72f810670a0562ea061461552840f5eb70
+      int count = 20;
+      while (count > 0) {
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: Suggested edit:

```diff
-      while (count > 0) {
-        if (RemoveDirectoryW(wpath)) {
-          break;
-        }
+      while (!RemoveDirectoryW(wpath) && count > 0) {
```

File: src/main/native/windows/file.cc [hunk 3]
```diff
@@ -446,20 +446,97 @@ int DeletePath(const wstring& path, wstring* error) {
...
           return DeletePathResult::kAccessDenied;
-        } else if (err == ERROR_DIR_NOT_EMPTY) {
-          // The directory is not empty.
-          return DeletePathResult::kDirectoryNotEmpty;
         } else if (err == ERROR_FILE_NOT_FOUND || err == ERROR_PATH_NOT_FOUND) {
           // The directory or one of its directories disappeared or is no longer
           // a directory.
           return DeletePathResult::kDoesNotExist;
+        } else if (err == ERROR_DIR_NOT_EMPTY) {
+          static const wstring kDot(L".");
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: Could you extract the file probing logic (lines 474..515) to a function?

Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Refactor done!

File: src/main/native/windows/file.cc (hunk 2 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/main/native/windows/file.cc (hunk 1 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/main/native/windows/file.cc [hunk 4]
```diff
@@ -393,6 +393,68 @@ int CreateJunction(const wstring& junction_name, const wstring& junction_target,
...
+struct DirectoryStatus {
+  enum {
+    kDoesNotExist = 0,
+    kDirectoryEmpty = 1,
+    kDirectoryNotEmpty = 2,
+    kPendingDeleteChildExists = 3,
+  };
+};
+
+// Check the status of a directory, the possible results are
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: This description is too generic.

How about: "Check whether the directory and its child elements truly exist, or are marked for deletion."

File: src/main/native/windows/file.cc [hunk 5]
```diff
@@ -471,72 +530,28 @@ int DeletePath(const wstring& path, wstring* error) {
...
-              if (found_pending_delete_file) {
-                Sleep(5L);
-              }
-              count--;
-              continue;
-            }
-          } else {
+          // We got ERROR_DIR_NOT_EMPTY error, but maybe the child files and dirs are already in deleting process,
+          // let's check the status of this directory to see if we should retry the delete operation.
+          int status = CheckDirectoryStatus(winpath);
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: How about:

```
switch (CheckDirectoryStatus(winpath)) {
  ...
}
```

File: src/main/native/windows/file.cc (hunk 4 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/main/native/windows/file.cc [hunk 6]
```diff
@@ -471,72 +530,28 @@ int DeletePath(const wstring& path, wstring* error) {
...
-              // delete again without waiting.
-              if (found_pending_delete_file) {
-                Sleep(5L);
-              }
-              count--;
-              continue;
-            }
-          } else {
+          // We got ERROR_DIR_NOT_EMPTY error, but maybe the child files and dirs are already in deleting process,
+          // let's check the status of this directory to see if we should retry the delete operation.
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Good point!

File: src/main/native/windows/file.cc (hunk 5 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done. Much cleaner!

**Comment by meteorcloudy** [comment]
Thanks for all the suggestion! PTAL~

File: src/main/native/windows/file.cc [hunk 7]
```diff
@@ -471,72 +530,28 @@ int DeletePath(const wstring& path, wstring* error) {
...
-              // the deleted files are gone, the directory is now empty, we can then try to
-              // delete again without waiting.
-              if (found_pending_delete_file) {
-                Sleep(5L);
-              }
-              count--;
-              continue;
-            }
-          } else {
+          // We got ERROR_DIR_NOT_EMPTY error, but maybe the child files and dirs are already in deleting process,
```
[the remark this answers is not in the document]
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/main/native/windows/file.cc [hunk 8]
```diff
@@ -393,6 +393,69 @@ int CreateJunction(const wstring& junction_name, const wstring& junction_target,
...
+        DWORD error_code = GetLastError();
+        // If the file or directory is in deleting process,
+        // CreateFileW returns ERROR_ACCESS_DENIED,
+        // If it's already deleted at the time we check,
+        // CreateFileW returns ERROR_FILE_NOT_FOUND.
+        // If CreateFileW fails with other reason, we consider there is a
+        // valid file that we cannot open, thus return kDirectoryNotEmpty
+        if (error_code != ERROR_ACCESS_DENIED &&
+            error_code != ERROR_FILE_NOT_FOUND) {
+          return DirectoryStatus::kDirectoryNotEmpty;
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: You also have to `FindClose` before returning. Maybe store the return value and just `break` out of the loop?

File: src/main/native/windows/file.cc [hunk 9]
```diff
@@ -393,6 +393,69 @@ int CreateJunction(const wstring& junction_name, const wstring& junction_target,
...
+        FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_FLAG_BACKUP_SEMANTICS,
+        NULL);
+      if (hFile != INVALID_HANDLE_VALUE) {
+        // If there is a valid file under the directory, the directory is truely not empty.
+        // We should just return kDirectoryNotEmpty.
+        CloseHandle(hFile);
+        return DirectoryStatus::kDirectoryNotEmpty;
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: You also have to `FindClose` before returning. Maybe store the return value and just `break` out of the loop?

File: src/main/native/windows/file.cc [hunk 10]
```diff
@@ -393,6 +393,69 @@ int CreateJunction(const wstring& junction_name, const wstring& junction_target,
...
+// 3. The directory contains valid files or dirs, so not empty
+// 4. The directory contains only files or dirs marked for deletion.
+int CheckDirectoryStatus(const wstring& path) {
+  static const wstring kDot(L".");
+  static const wstring kDotDot(L"..");
+  bool found_child_marked_for_deletion = false;
+  WIN32_FIND_DATAW metadata;
+  HANDLE handle = ::FindFirstFileW((path + L"\\*").c_str(), &metadata);
+  if (handle == INVALID_HANDLE_VALUE) {
+    return DirectoryStatus::kDoesNotExist;;
```
Comment (laszlocsomor) [inline src/main/native/windows/file.cc]: nit: double ;;

File: src/java_tools/buildjar/java/com/google/devtools/build/buildjar/SimpleJavaLibraryBuilder.java [hunk 11]
```diff
@@ -211,6 +211,7 @@ public FileVisitResult visitFile(Path path, BasicFileAttributes attrs)
     if (protobufMetadataBuffer.size() > 0) {
       try (OutputStream outputStream = Files.newOutputStream(output)) {
         protobufMetadataBuffer.writeTo(outputStream);
+        outputStream.close();
```
Comment (cushon) [inline src/java_tools/buildjar/java/com/google/devtools/build/buildjar/SimpleJavaLibraryBuilder.java]: The try-with-resources should close the output stream, why is this necessary?

File: src/main/native/windows/file.cc (hunk 8 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done. I just reverted back to the way that uses `find_valid_file`

File: src/main/native/windows/file.cc (hunk 9 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/main/native/windows/file.cc (hunk 10 above)
Comment (meteorcloudy) [reply src/main/native/windows/file.cc]: Done.

File: src/java_tools/buildjar/java/com/google/devtools/build/buildjar/SimpleJavaLibraryBuilder.java (hunk 11 above)
Comment (meteorcloudy) [reply src/java_tools/buildjar/java/com/google/devtools/build/buildjar/SimpleJavaLibraryBuilder.java]: Oh, sorry I didn't mean to make this change. I was just debugging the Java build failure, but yes, looks like this should not be necessary.

**Comment by meteorcloudy** [comment]
Please take a look again, thanks!
